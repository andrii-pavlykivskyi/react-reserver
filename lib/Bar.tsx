import React, { ReactNode, ComponentProps, ReactElement } from 'react';
import { Dimension } from './types';

type Content = ReactNode & { props: { style: React.CSSProperties } };

function getContent(
  index: number,
  length: number,
  firstContent: ReactNode,
  lastContent: ReactNode
): Content {
  if (index === 0) {
    return (firstContent || <div />) as Content;
  } else if (length - 1 === index) {
    return (lastContent || <div />) as Content;
  }
  return <div />;
}

export type BarProps = Omit<ComponentProps<'div'>, 'content'> & {
  length: number;
  dimension: Dimension;
  firstContent: ReactNode;
  lastContent: ReactNode;
};

export default function Bar({
  length = 1,
  dimension,
  firstContent = null,
  lastContent = null,
  ...props
}: BarProps) {
  return (
    <div
      {...props}
      id={props.id}
      role="listitem"
      draggable={props.draggable}
      style={{
        ...props.style,
        pointerEvents: props.style?.pointerEvents || 'none',
        background: props.style?.background || '#0E6BA8',
        display: props.style?.display || 'flex',
        position: props.style?.position || 'absolute',
        zIndex: props.style?.zIndex || '100',
        cursor: 'move'
      }}
      className={props.className}
    >
      {[...Array(length)].map((_, i) => {
        const processedContent = getContent(i, length, firstContent, lastContent);

        const style = Object.assign(
          {
            width: dimension.width,
            height: dimension.height,
            pointerEvents: props.style?.pointerEvents || 'none'
          },
          (processedContent?.props && processedContent?.props?.style) || {}
        );

        return (
          <React.Fragment key={i}>
            {React.cloneElement(
              processedContent as ReactElement,
              { ...processedContent.props, style },
              processedContent.props.children
            )}
          </React.Fragment>
        );
      })}
      {props.children}
    </div>
  );
}
