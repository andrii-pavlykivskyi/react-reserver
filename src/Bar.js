import React from 'react'

export default function Bar(props) {

  return (
    <div
      role='listitem'
      onDragStart={props.onDragStart}
      onClick={(e) => {
        props.onClick(e, props)
      }}
      onMouseOver={(e) => {
        props.onMouseOver(e, props)
      }}
      onContextMenu={(e) => {        
        props.onContextMenu(e,props)
      }}
      onMouseEnter={(e) => {        
        props.onMouseEnter(e,props)
      }}
       onMouseLeave={(e) => {        
        props.onMouseLeave(e,props)
      }}
      draggable={props.draggable}
      style={{
        ...props.style,
        pointerEvents: props.style.pointerEvents || 'none',
        background: props.style.background || '#0E6BA8',
        display: 'flex',
        position: 'absolute',
        zIndex: props.zIndex || '100'
      }}
      className={props.className}
    >
      {[...Array(props.length && props.length > 0 ? props.length : 1)].map(
        (notUsed, i) => {
          return (
            <div
              key={i}
              style={{
                width: props.dimension,
                height: props.dimension,
                pointerEvents:  props.style.pointerEvents || 'none'
              }}
            />
          )
        }
      )}
      {props.children}
    </div>
  )
}

Bar.defaultProps = {
  style: {},
  dimension: 20,
  onClick: () => {},
  onMouseOver: () => {},
  onDragStart: () => {},
  onContextMenu: () => {},
  onMouseEnter:()=>{},
  onMouseLeave:()=>{},
  length:1
}
