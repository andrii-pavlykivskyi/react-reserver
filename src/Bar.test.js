import React from 'react'
import Bar from './Bar'
import { render, fireEvent, screen } from '@testing-library/react'
import {
  toBeInTheDocument,
  toHaveStyle
} from '@testing-library/jest-dom/matchers'
import renderer from 'react-test-renderer'

expect.extend({ toBeInTheDocument, toHaveStyle })

test('Bar props changes color ', () => {
  render(<Bar dimension={{ width: 20, height: 20 }} background='red' />)

  expect(screen.getByRole('listitem')).toHaveStyle(
    'background:rgb(14, 107, 168)'
  )

  // screen.getByText('1')
})

test('onClick event fires', () => {
  const onClick = jest.fn()
  render(<Bar dimension={{ width: 20, height: 20 }} onClick={onClick} />)
  fireEvent.click(screen.getByRole('listitem'))

  expect(onClick).toHaveBeenCalled()
})

test('onContextMenu event fires', () => {
  const onClick = jest.fn()
  render(
    <Bar
      style={{ pointerEvents: 'auto' }}
      dimension={{ width: 20, height: 20 }}
      onContextMenu={onClick}
    />
  )
  fireEvent.contextMenu(screen.getByRole('listitem'))

  expect(onClick).toHaveBeenCalled()
})

test('onContextMenu event fires', () => {
  const onClick = jest.fn()
  render(
    <Bar
      style={{ pointerEvents: 'auto' }}
      dimension={{ width: 20, height: 20 }}
      onPointerDown={onClick}
    />
  )
  fireEvent.pointerDown(screen.getByRole('listitem'))

  expect(onClick).toHaveBeenCalled()
})

test('onDragStart event fires', () => {
  const dragMe = jest.fn()
  render(
    <Bar
      style={{ pointerEvents: 'auto' }}
      dimension={{ width: 20, height: 20 }}
      onDragStart={dragMe}
    />
  )
  fireEvent.dragStart(screen.getByRole('listitem'))

  expect(dragMe).toHaveBeenCalled()
})

test('onDragEnd event fires', () => {
  const dragMe = jest.fn()
  render(
    <Bar
      style={{ pointerEvents: 'auto' }}
      dimension={{ width: 20, height: 20 }}
      onDragEnd={dragMe}
    />
  )
  fireEvent.dragEnd(screen.getByRole('listitem'))

  expect(dragMe).toHaveBeenCalled()
})

test('onMouseMove event fires', () => {
  const dragMe = jest.fn()
  render(
    <Bar
      style={{ pointerEvents: 'auto' }}
      dimension={{ width: 20, height: 20 }}
      onMouseMove={dragMe}
    />
  )
  fireEvent.mouseMove(screen.getByRole('listitem'))

  expect(dragMe).toHaveBeenCalled()
})

test('onMouseDown event fires', () => {
  const dragMe = jest.fn()
  render(
    <Bar
      style={{ pointerEvents: 'auto' }}
      dimension={{ width: 20, height: 20 }}
      onMouseDown={dragMe}
    />
  )
  fireEvent.mouseDown(screen.getByRole('listitem'))

  expect(dragMe).toHaveBeenCalled()
})

test('onMouseUp event fires', () => {
  const dragMe = jest.fn()
  render(
    <Bar
      style={{ pointerEvents: 'auto' }}
      dimension={{ width: 20, height: 20 }}
      onMouseUp={dragMe}
    />
  )
  fireEvent.mouseUp(screen.getByRole('listitem'))

  expect(dragMe).toHaveBeenCalled()
})
test('onContextMenu event fires', () => {
  const onClick = jest.fn()
  render(
    <Bar
      style={{ pointerEvents: 'auto' }}
      dimension={{ width: 20, height: 20 }}
      onContextMenu={onClick}
    />
  )
  fireEvent.contextMenu(screen.getByRole('listitem'))

  expect(onClick).toHaveBeenCalled()
})

test('onMouseOver event fires', () => {
  const onClick = jest.fn()
  render(
    <Bar
      style={{ pointerEvents: 'auto' }}
      dimension={{ width: 20, height: 20 }}
      onMouseOver={onClick}
    />
  )
  fireEvent.mouseOver(screen.getByRole('listitem'))

  expect(onClick).toHaveBeenCalled()
})

test('should have default onDragStart', () => {
  expect(Bar.defaultProps.onDragStart).toBeDefined()
})

test('should have default onClick', () => {
  expect(Bar.defaultProps.onClick).toBeDefined()
})

test('renders content', () => {
  const myComp = (props) => (
    <div style={{ background: 'green' }}>{props.children || 'compy'} </div>
  )
  const bar = renderer
    .create(
      <Bar
        style={{ pointerEvents: 'auto' }}
        dimension={{ width: 20, height: 20 }}
        length={4}
        content={{ 2: <div>cool</div> }}
        firstContent={<myComp />}
        lastContent={<myComp>hobo</myComp>}
      />
    )
    .toJSON()
  expect(bar).toMatchSnapshot()
})

test('renders content only body', () => {
  const bar = renderer
    .create(
      <Bar
        style={{ pointerEvents: 'auto' }}
        dimension={{ width: 20, height: 20 }}
        length={4}
        content={{ 0: <div>great</div>, 2: <div>cool</div>, 3: <div>meh</div> }}
      />
    )
    .toJSON()
  expect(bar).toMatchSnapshot()
})

test('renders content only last ', () => {
  const bar = renderer
    .create(
      <Bar
        style={{ pointerEvents: 'auto' }}
        dimension={{ width: 20, height: 20 }}
        length={4}
        content={{ 2: <div>horror</div> }}
      />
    )
    .toJSON()
  expect(bar).toMatchSnapshot()
})

test('make sure default prop didnt change', () => {
  expect(Bar.defaultProps.style).toMatchObject({})
  expect(Bar.defaultProps.content).toMatchObject({})
  expect(Bar.defaultProps.dimension).toMatchObject({ width: 20, height: 20 })
  expect(Bar.defaultProps.length).toBe(1)
  ;[
    'onClick',
    'onDragStart',
    'onDragEnd',
    'onMouseOver',
    'onContextMenu',
    'onMouseDown',
    'onMouseUp',
    'onMouseEnter',
    'onMouseLeave',
    'onMouseMove'
  ].forEach((name) => {
    const func = jest.spyOn(Bar.defaultProps, name)
    Bar.defaultProps[name]()
    expect(func).toHaveBeenCalled()
  })
})
