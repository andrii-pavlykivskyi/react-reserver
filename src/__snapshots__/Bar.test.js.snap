// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders content 1`] = `
<div
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragEnd={[Function]}
  onDragStart={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseMove={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  role="listitem"
  style={
    Object {
      "background": "#0E6BA8",
      "display": "flex",
      "pointerEvents": "auto",
      "position": "absolute",
      "zIndex": "100",
    }
  }
>
  <myComp
    style={
      Object {
        "height": 20,
        "pointerEvents": "auto",
        "width": 20,
      }
    }
  />
  <div
    style={
      Object {
        "height": 20,
        "pointerEvents": "auto",
        "width": 20,
      }
    }
  />
  <div
    style={
      Object {
        "height": 20,
        "pointerEvents": "auto",
        "width": 20,
      }
    }
  />
  <myComp
    style={
      Object {
        "height": 20,
        "pointerEvents": "auto",
        "width": 20,
      }
    }
  >
    hobo
  </myComp>
</div>
`;

exports[`renders content only body 1`] = `
<div
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragEnd={[Function]}
  onDragStart={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseMove={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  role="listitem"
  style={
    Object {
      "background": "#0E6BA8",
      "display": "flex",
      "pointerEvents": "auto",
      "position": "absolute",
      "zIndex": "100",
    }
  }
>
  <div
    style={
      Object {
        "height": 20,
        "pointerEvents": "auto",
        "width": 20,
      }
    }
  />
  <div
    style={
      Object {
        "height": 20,
        "pointerEvents": "auto",
        "width": 20,
      }
    }
  />
  <div
    style={
      Object {
        "height": 20,
        "pointerEvents": "auto",
        "width": 20,
      }
    }
  />
  <div
    style={
      Object {
        "height": 20,
        "pointerEvents": "auto",
        "width": 20,
      }
    }
  />
</div>
`;

exports[`renders content only last  1`] = `
<div
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragEnd={[Function]}
  onDragStart={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseMove={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  role="listitem"
  style={
    Object {
      "background": "#0E6BA8",
      "display": "flex",
      "pointerEvents": "auto",
      "position": "absolute",
      "zIndex": "100",
    }
  }
>
  <div
    style={
      Object {
        "height": 20,
        "pointerEvents": "auto",
        "width": 20,
      }
    }
  />
  <div
    style={
      Object {
        "height": 20,
        "pointerEvents": "auto",
        "width": 20,
      }
    }
  />
  <div
    style={
      Object {
        "height": 20,
        "pointerEvents": "auto",
        "width": 20,
      }
    }
  />
  <div
    style={
      Object {
        "height": 20,
        "pointerEvents": "auto",
        "width": 20,
      }
    }
  />
</div>
`;
